HW5.py

```diff
@@ -0,0 +1,586 @@
+    file_name = f"plots/{pred_name}-plot.html"
+    fig = go.Figure()
+    for curr_hist, curr_group in zip(pred_data, response):
+        fig.add_trace(
+            go.Violin(
+                x=np.repeat(curr_group, 200),
+                y=[curr_hist],
+                name=curr_group,
+                box_visible=True,
+                meanline_visible=True,
+            )
+        )
+    fig.update_layout(
+        title="Continuous Response by Categorical Predictor",
+        xaxis_title=pred_name,
+        yaxis_title="Response",
+    )
+    fig.write_html(
+        file=file_name,
+        include_plotlyjs="cdn",
+        include_mathjax="cdn",
+    )
+
+    return file_name
+
+
+def plot_cat_resp_con_predictor(pred_data, pred_name, response):
+    file_name = f"plots/{pred_name}-plot.html"
+    group_labels = list(set([int(x) for x in response]))
+    hist_data0 = pred_data[response == group_labels[0]]
+    hist_data1 = pred_data[response == group_labels[1]]
+    hist_data = [hist_data0, hist_data1]
+    fig = ff.create_distplot(hist_data, group_labels, bin_size=0.2)
+    fig.update_layout(
+        title="Continuous Predictor by Categorical Response",
+        xaxis_title=pred_name,
+        yaxis_title="Distribution",
+    )
+
+    fig.write_html(
+        file=file_name,
+        include_plotlyjs="cdn",
+        include_mathjax="cdn",
+    )
+
+    return file_name
+
+
+def plot_con_response_con_predictor(pred_data, pred_name, response):
+    file_name = f"plots/{pred_name}-plot.html"
+    fig = px.scatter(x=pred_data, y=response, trendline="ols")
+    fig.update_layout(
+        title="Continuous Response by Continuous Predictor",
+        xaxis_title=pred_name,
+        yaxis_title="Response",
+    )
+
+    fig.write_html(
+        file=file_name,
+        include_plotlyjs="cdn",
+        include_mathjax="cdn",
+    )
+
+    return file_name
+
+
+def con_con_correlation(pred1_data, pred1_name, pred2_data, pred2_name):
+    correlation_stats, _ = scipy.stats.pearsonr(pred1_data, pred2_data)
+    return correlation_stats
+
+
+def con_cat_correlation(
+    pred1_data, pred1_name, pred1_type, pred2_data, pred2_name, pred2_type
+):
+    if len(set(pred1_data)) == 2 or len(set(pred2_data)) == 2:
+        pred1_data, _ = pd.factorize(pred1_data)
+        correlation_stats, _ = scipy.stats.pointbiserialr(pred1_data, pred2_data)
+    elif pred1_type == "cat":
+        correlation_stats = cat_cont_correlation_ratio(
+            np.array(pred1_data), np.array(pred2_data)
+        )
+    elif pred2_type == "cat":
+        correlation_stats = cat_cont_correlation_ratio(
+            np.array(pred2_data), np.array(pred1_data)
+        )
+    return correlation_stats
+
+
+def cat_cat_correlation(pred1_data, pred1_name, pred2_data, pred2_name):
+    correlation_stats = cat_correlation(pred1_data, pred2_data)
+    return correlation_stats
+
+
+def plot_difference_with_mean_unweighted(
```

Comment:
Adding annotations (specifically population counts) might help make your weighted and unweighted difference plots easier to use (specifically, it helps show how big/small a given bin is).